### PR TITLE
[8.x] Job Batching

### PR DESCRIPTION
**Pull request by taylorotwell**
This PR implements job batching:

### Dispatching Batches

image

The `Bus::batch` method returns an `Illuminate\Bus\Batch` instance which may be inspected to gather the ID and other information about the batch. The batch instance is also JSON serializable so it may be returned directly from a route:

image

When dispatching batches, you may attach various callbacks to the batch using `then`, `success`, and `catch`. 

In addition, you may use the `allowFailures` method to indicate that a batch should not be automatically cancelled when a job within the batch fails. If this method is not called, the batch will automatically be cancelled when a job within the batch fails.

### Querying Batches

Batch meta-data storage is backend agnostic behind a `BatchRepository` interface. I have implemented a database implementation. You may retrieve an existing batch using the `Bus::findBatch` method, which either returns an `Illuminate\Bus\Batch` instance of `null`:

image

The `Batch` instance has various helper methods on it, such as `finished`, `cancelled`, `cancel`, etc.

### Batchable Jobs

Batchable jobs should use a new `Batchable` trait:

image

At the beginning of a batched job's `handle` method, it may detect if the batch has been cancelled and act accordingly:

image

### Adding Additional Jobs To A Batch

Within a batched job, you may add additional jobs to the batch using the `add` method on the batch instance. You **should never** add jobs to an existing batch outside of a batched job:

image

### Failed Job Handling

The UUIDs of the failed jobs within a batch are stored with the batch meta data. A new `queue:retry-batch {batch-uuid}` command has been added to easily retry all of the failed jobs for a given batch. **Note that you can only retry jobs for a batch that allows failures.** 

image

> Supporting this feature required the addition of a new failed job backend implementation for relational databases that uses the job's UUID instead of an auto-incrementing primary key to reference jobs. I decided to make this a new implementation (which will be the default in Laravel 8.x) since it would be easier to allow users to continue using the existing version until they actually need batches.

### Testing

You may test the dispatching of batches using `Bus::fake` and `Bus::assertBatched`:

image

### CODE REVIEW

File: src/Illuminate/Support/Testing/Fakes/BusFake.php [hunk 1]
```diff
@@ -320,6 +360,40 @@ public function dispatchAfterResponse($command)
...
+     */
+    public function batch($jobs)
+    {
+        return new PendingBatchFake($this, Collection::wrap($jobs));
+    }
+
+    /**
+     * Record the fake pending batch dispatch.
+     *
+     * @param  \Illuminate\Bus\PendingBatch
```
Comment (voyula) [inline src/Illuminate/Support/Testing/Fakes/BusFake.php:387]: Incorrect docblock.

File: src/Illuminate/Support/Testing/Fakes/BusFake.php [hunk 2]
```diff
@@ -320,6 +360,40 @@ public function dispatchAfterResponse($command)
         }
     }
 
+    /**
+     * Attempt to find the batch with the given ID.
+     *
+     * @param  string  $batchId
+     * @return \Illuminate\Bus\Batch|null
+     */
+    public function findBatch(string $batchId)
```
Comment (voyula) [inline src/Illuminate/Support/Testing/Fakes/BusFake.php:369]: Type hinting?

File: src/Illuminate/Bus/PendingBatch.php [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+     */
+    public function connection()
+    {
+        return $this->options['connection'] ?? null;
+    }
+
+    /**
+     * Specify the queue that the batched jobs should run on.
+     *
+     * @param  string  $connection
```
Comment (voyula) [inline src/Illuminate/Bus/PendingBatch.php:159]: Incorrect parameter name.

File: src/Illuminate/Bus/Batch.php [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+     * The date indicating when the batch was finished.
+     *
+     * @var \Illuminate\Support\CarbonImmutable|null
+     */
+    public $finishedAt;
+
+    /**
+     * Create a new batch instance.
+     *
+     * @param  \Illuminate\Contracts\Bus\Dispatcher  $bus
```
Comment (voyula) [inline src/Illuminate/Bus/Batch.php:93]: Incorrect parameter name and type.

File: src/Illuminate/Bus/Batch.php [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+     * @param  \Illuminate\Bus\BatchRepository  $repository
+     * @param  string  $id
+     * @param  string  $id
```
Comment (voyula) [inline src/Illuminate/Bus/Batch.php:96]: Duplicated @param.

File: src/Illuminate/Queue/Console/RetryBatchCommand.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        $batch = $this->laravel[BatchRepository::class]->find($this->argument('id'));
+
+        if (! $batch) {
+            $this->error("Unable to find a batch with ID [{$id}].");
+
+            return 1;
+        }
+
+        if (empty($batch->failedJobIds)) {
+            $this->error("The given batch does not contain any failed jobs.");
```
Comment (voyula) [inline src/Illuminate/Queue/Console/RetryBatchCommand.php]: Can make single quote?

File: src/Illuminate/Support/Testing/Fakes/BusFake.php (hunk 1 above)
Comment (taylorotwell) [reply src/Illuminate/Support/Testing/Fakes/BusFake.php:387]: How?

**Comment by mattstauffer** [comment]
Oh my lord you don't know how excited I am for this!!

**Comment by fgilio** [comment]
Haha that makes two of us

**Comment by DarkGhostHunter** [comment]
Wh-- Why autoincrementing IDs can't be used here?

File: src/Illuminate/Support/Testing/Fakes/BusFake.php (hunk 1 above)
Comment (voyula) [reply src/Illuminate/Support/Testing/Fakes/BusFake.php:387]: @taylorotwell Haven't parameter name $pendingBatch.

Comment (voyula) [inline src/Illuminate/Support/Testing/Fakes/BusFake.php:387]: Haven't parameter name $pendingBatch.

@param  \Illuminate\Bus\PendingBatch $pendingBatch

**Comment by SjorsO** [comment]
Looks great!

Small nitpick, but maybe you should go with `canceled` (with one L). The spelling with one L is more common in american English, with 2 L's is more common in British English.

**Comment by taylorotwell** [comment]
@SjorsO I'll just add an alias so either will work.

File: src/Illuminate/Queue/Console/RetryBatchCommand.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Retry the failed jobs for a batch';
+
+    /**
+     * Execute the console command.
+     *
+     * @return void
```
Comment (voyula) [inline src/Illuminate/Queue/Console/RetryBatchCommand.php:27]: Incorrect return type.

Should be;
@return int|null

**Review by vishal-adhikari — APPROVED** [review]

File: src/Illuminate/Bus/Batch.php [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    public function __construct(QueueFactory $queue,
+                                BatchRepository $repository,
+                                string $id,
+                                int $totalJobs,
+                                int $pendingJobs,
+                                int $failedJobs,
+                                array $failedJobIds,
+                                array $options,
+                                CarbonImmutable $createdAt,
+                                ?CarbonImmutable $cancelledAt = null,
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/Batch.php:115]: Usually laravel does not use the questionmark and implements nullable types with the default value being null. Not both.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Bus/Batch.php (hunk 8 above)
Comment (X-Coder264) [reply src/Illuminate/Bus/Batch.php:115]: That is legacy PHP behavior which IMO shouldn't be used anymore. The new database factory code also uses the question mark.

Comment (X-Coder264) [reply src/Illuminate/Bus/Batch.php:115]: https://wiki.php.net/rfc/typed_properties_v2

> This is in contrast to parameter types, where a null default value automatically implies a nullable type. We consider this to be a legacy behavior, which we do not wish to support for newly introduced syntax.

Comment (GrahamCampbell) [reply src/Illuminate/Bus/Batch.php:115]: Many people still use the style without the question mark. Including me.

Comment (GrahamCampbell) [reply src/Illuminate/Bus/Batch.php:115]: (and the rest of the laravel/framework source code)

Comment (GrahamCampbell) [reply src/Illuminate/Bus/Batch.php:115]: If laravel wants to change its code style to use the question mark, StyleCI can enforce this...

Comment (GrahamCampbell) [reply src/Illuminate/Bus/Batch.php:115]: It can also enforce non-usage if we want to enable that rule on the laravel preset.

**Comment by rasmuscnielsen** [comment]
Looks awesome 🎉

Was wondering what happens when a batch gets cancelled? 
Would expect the remaining pending jobs to get cancelled, but from the example it seems that this has to be implemented manually in the individual jobs' handle() methods?

**Comment by SjorsO** [comment]
> @SjorsO I'll just add an alias so either will work.

I think consistent American English spelling throughout the framework would be a better idea. That way teams don't have to bikeshed about canceled/cancelled, color/colour, etc

File: src/Illuminate/Bus/Batch.php [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+                $data = '',
+                $this->options['queue'] ?? null
+            );
+        });
+
+        return $this->fresh();
+    }
+
+    /**
+     * Get the total number of jobs that have been processed by the batch thus far.
```
Comment (voyula) [inline src/Illuminate/Bus/Batch.php:167]: Where is the return type?

File: src/Illuminate/Bus/DatabaseBatchRepository.php [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+        $this->factory = $factory;
+        $this->connection = $connection;
+        $this->table = $table;
+    }
+
+    /**
+     * Retrieve information about an existing batch.
+     *
+     * @param  string  $batchId
+     * @return \Illuminate\Bus\Batch
```
Comment (voyula) [inline src/Illuminate/Bus/DatabaseBatchRepository.php:52]: Should be;

@return \Illuminate\Bus\Batch|null

**Comment by themsaid** [comment]
@rasmuscnielsen that's right. Before each job you need to check if the batch is still active before you attempt to run the code.

**Comment by anouarabdsslm** [comment]
Waiting for this to get merged to play with it. thanks guys for this cool feature

**Comment by taylorotwell** [comment]
@rasmuscnielsen that is correct. that is somewhat modeled after Sidekiq Pro's approach in Ruby. It is the most flexible approach. We could later add helpers on the batch itself to make this more automatic...

`Bus::batch(...)->failCancelled()` 

Or something to that effect...

**Comment by ahmedash95** [comment]
Nice feature, Is there any way to get the failed jobs instead of the number of failers?

```php
<?php

->catch(function(Batch $batch, $e){
       $batch->failedJobs->each(function($job){
             info(get_class($job) .' faild with exception: '. $e->getMessage());
      })
})
```

**Comment by themsaid** [comment]
@ahmedash95 there's a `$failedJobIds`public property on the Batch instance that you can use.

**Comment by ahmedash95** [comment]
> @ahmedash95 there's a `$failedJobIds`public property on the Batch instance that you can use.

Prefect 👌🙏

**Comment by taylorotwell** [comment]
Note that you would only be able to get the full list of `failedJobIds` from the `then` callback though.

**Comment by taylorotwell** [comment]
Noting that since this was merged... `success` has been renamed to `then`. What was `then` is now `finally`. This better matches JS callback terminology.

File: src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    {
+    }
+
+    /**
+     * Decrement the total number of pending jobs for the batch.
+     *
+     * @param  string  $batchId
+     * @return int|null
+     */
+    public function decrementPendingJobs(string $batchId)
```
Comment (mfn) [inline src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php:64]: This is `BatchRepositoryFake implements BatchRepository` but the method on the interface is `public function decrementPendingJobs(string $batchId, string $jobId);`, as such this doesn't work due to:
> `PHP Fatal error:  Declaration of Illuminate\Support\Testing\Fakes\BatchRepositoryFake::decrementPendingJobs(string $batchId) must be compatible with Illuminate\Bus\BatchRepository::decrementPendingJobs(string $batchId, string $jobId) in …/src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php on line …`

File: src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        return 0;
+    }
+
+    /**
+     * Increment the total number of failed jobs for the batch.
+     *
+     * @param  string  $batchId
+     * @return int|null
+     */
+    public function incrementFailedJobs(string $batchId)
```
Comment (mfn) [inline src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php:75]: Similar error as above, interface method is `public function incrementFailedJobs(string $batchId, string $jobId);`
> `PHP Fatal error:  Declaration of Illuminate\Support\Testing\Fakes\BatchRepositoryFake::incrementFailedJobs(string $batchId) must be compatible with Illuminate\Bus\BatchRepository::incrementFailedJobs(string $batchId, string $jobId) in …src/Illuminate/Support/Testing/Fakes/BatchRepositoryFake.php on line …`

**Review by mfn — COMMENTED** [review]
This probably wasn't caught as there being no tests for the fake.

I got this when I ran parts of laravel test suites with pcov code coverage, which I guess triggered loading of this class and then failed.

**Pull request merged**